app/src/main/java/co/tinode/tindroid/db/MessageDb.java

```diff
@@ -301,6 +301,53 @@ static Cursor queryDeleted(SQLiteDatabase db, long topicId, boolean hard) {
         return db.rawQuery(sql, null);
     }
 
+    /**
+     * Find the latest missing range of messages for fetching from the server.
+     *
+     * @param db database to select from;
+     * @param topicId Tinode topic ID (topics._id) to select from;
+     * @return range of missing IDs if found null if either all messages are present or no messages are found.
+     */
+    static MsgRange getNextMissingRange(SQLiteDatabase db, long topicId) {
+        int high = 0;
+        final String sqlHigh = "SELECT MAX(m1." + COLUMN_NAME_SEQ + ") AS missing" +
+                " FROM " + TABLE_NAME + " AS m1" +
+                " LEFT JOIN " + TABLE_NAME + " AS m2" +
+                " ON m1." + COLUMN_NAME_SEQ + "=IFNULL(m2." + COLUMN_NAME_HIGH + ", m2." + COLUMN_NAME_SEQ + "+1)" +
```

Comment:
`IFNULL(.., m2.COLUMN_NAME_SEQ + 1)` will return NULL given line 318.

Reply:
I'm not sure about this comment. I tested the query on actual data. It seems to work correctly.

Reply:
Since m2.COLUMN_NAME_SEQ is always NULL
`IFNULL(m2.COLUMN_NAME_HIGH, m2.COLUMN_NAME_SEQ + 1)` is redundant.
I think I understand this query now, though. No changes needed.

Reply:
No, it's not always NULL. This condition is what the LEFT JOIN is being joined on. Just try this query either on MYSQL or on sqlite. The SQL here is a bit hairy but I believe it's correct.